### ISSUE TITLE
TypeError: Cannot read property 'name' of undefined

### ISSUE THREAD

**Issue by jpollard-cs** [RobinBuschmann/sequelize-typescript#712]
**Versions**

- sequelize: ^5.17.2
- sequelize-typescript: 1.0.0
- typescript: 3.7.2

**I'm submitting a ...** 
[x] bug report
[ ] feature request

**Actual behavior:**
```
    TypeError: Cannot read property 'name' of undefined
      at Object.getForeignKeyOptions (node_modules/sequelize-typescript/dist/associations/foreign-key/foreign-key-service.js:24:40)
      at HasAssociation.getSequelizeOptions (node_modules/sequelize-typescript/dist/associations/has/has-association.js:17:52)
      at associations.forEach.association (node_modules/sequelize-typescript/dist/sequelize/sequelize/sequelize.js:49:45)
          at Array.forEach (<anonymous>)
      at models.forEach.model (node_modules/sequelize-typescript/dist/sequelize/sequelize/sequelize.js:48:26)
          at Array.forEach (<anonymous>)
      at Sequelize.associateModels (node_modules/sequelize-typescript/dist/sequelize/sequelize/sequelize.js:44:16)
      at Sequelize.addModels (node_modules/sequelize-typescript/dist/sequelize/sequelize/sequelize.js:36:14)
      at new Sequelize (node_modules/sequelize-typescript/dist/sequelize/sequelize/sequelize.js:18:22)
      at SequelizeConnection.initialize (src/common/infrastructure/sequelize.ts:30:21)
      at DatabaseTestFrameworkSimple.<anonymous> (test/framework/database.ts:74:24)
      at test/framework/database.ts:8:71
      at Promise._execute (node_modules/bluebird/js/release/debuggability.js:427:9)
      at Promise._resolveFromExecutor (node_modules/bluebird/js/release/promise.js:518:18)
      at new Promise (node_modules/bluebird/js/release/promise.js:103:10)
      at Object.<anonymous>.__awaiter (test/framework/database.ts:4:12)
      at DatabaseTestFrameworkSimple.initializeTestDb (test/framework/database.ts:71:16)
      at DatabaseTestFrameworkSimple.<anonymous> (test/framework/database.ts:98:16)
      at fulfilled (test/framework/database.ts:5:58)
      at tryCatcher (node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (node_modules/bluebird/js/release/promise.js:547:31)
      at Promise._settlePromise (node_modules/bluebird/js/release/promise.js:604:18)
      at Promise._settlePromise0 (node_modules/bluebird/js/release/promise.js:649:10)
      at Promise._settlePromises (node_modules/bluebird/js/release/promise.js:729:18)
      at _drainQueueStep (node_modules/bluebird/js/release/async.js:93:12)
      at _drainQueue (node_modules/bluebird/js/release/async.js:86:9)
      at Async.Object.<anonymous>.Async._drainQueues (node_modules/bluebird/js/release/async.js:102:5)
      at Immediate.Async.drainQueues [as _onImmediate] (node_modules/bluebird/js/release/async.js:15:14)
```

**Related code:**

I have a many-to-many relationship which is essentially modeled like is seen in the below code. The `VendorEmployeeModel` represents a join table, but there may not always be an `Employee` record unless they sign up and link their employee record to their `VendorEmployee` record  (the vendor may just create a `VendorEmployee` record to keep track of what each `Employee` is currently working on). The intent is to be able to `include` and query on the `VendorEmployeeModel` which I am unclear if this can be done with `BelongsToMany`. I wasn't clear why this is failing because I am using the `BelongsTo` and `HasMany` decorators throughout the codebase. The only thing that seems to be different here is a `Vendor` can have many `VendorEmployee`s *and* a `VendorEmployee` can have many `Vendors`. Besides the main error I'm also seeing a `SequelizeUniqueConstraintError: Validation error` and `TypeError: Cannot read property 'query' of undefined`. If I remove the `VendorEmployee` and `Vendor` `BelongsTo / HasMany` decorators then everything works fine (other the `includes` in my repository). Apologies if this is not a sequelize-typescript specific issue, but any guidance here would be greatly appreciated.

```ts
@Table({
  tableName: 'VendorEmployee',
  timestamps: true
})
export default class VendorEmployeeModel extends Model<VendorEmployeeModel> {
  @Default(() => IdUtility.generateUid())
  @Column
  uid!: string;

  @ForeignKey(() => VendorModel)
  @Column
  vendorId!: number;

  @BelongsTo(() => VendorModel)
  vendor: VendorModel;

  @ForeignKey(() => EmployeeModel)
  @Column
  employeeId: number;

  @BelongsTo(() => EmployeeModel)
  employee: EmployeeModel;

  @IsEmail
  @Column
  email: string;

  @Column
  phone!: string;

  @Column
  firstName!: string;

  @Column
  lastName!: string;
}

@Table({
  tableName: 'Vendor',
  timestamps: true
})
export default class VendorModel extends Model<VendorModel> {
  @Unique
  @Default(() => IdUtility.generateUid())
  @Column
  uid: string;

  @HasMany(() => VendorEmployeeModel)
  vendorEmployees: VendorEmployeeModel[];
}

@Table({
  tableName: 'Employee'
})
export default class EmployeeModel extends Model<EmployeeModel> {
  @Unique
  @Default(() => IdUtility.generateUid())
  @Column
  uid: string;

  @Unique
  @Column
  userUid: string;

  @HasMany(() => VendorEmployeeModel)
  carrierDriver: VendorEmployeeModel[];
}
```

**Comment by jpollard-cs** [comment]
I've put this up on StackOverflow with some improved formatting: https://stackoverflow.com/questions/59195662/sequelize-many-to-many-relationship-typeerror-cannot-read-property-name-of-un

**Comment by RobinBuschmann** [comment]
Does your mentioned code also reflect the file structure. What I mean is, are your classes in one file? Can you also provide how you initialize the sequelize instance?

**Comment by jpollard-cs** [comment]
@RobinBuschmann these are in separate files
Will follow up in a moment with the initialization code

**Comment by jpollard-cs** [comment]
so we have a custom wrapper that looks something like this:

```
export class SequelizeConnection {
  private readonly uri: string;
  private readonly sequelizeConfig: SequelizeOptions;
  private instance: Sequelize;

  constructor(uri: string, options?: SequelizeOptions) {
    this.uri = uri;
    this.sequelizeConfig = options;
  }

  public initialize(overrideUri?: string) {
    log.info('Initialize sequelize connection');
    this.instance = new Sequelize(overrideUri ? overrideUri : this.uri, this.sequelizeConfig);
  }

  public getInstance() {
    return this.instance;
  }

  public async checkDatabasePresence(name: string) {
    if (!this.instance) {
      log.warn('Sequelize connection must be initialized before checking for database presence');
      return;
    }

    log.info(`Checking for presence of database: ${name}`);
    try {
      const result = await this.instance.query(`SELECT 1 FROM pg_database WHERE datname='${name}'`);
      if (result[0].length > 0) {
        log.info(`Database '${name}' exists`);
      } else {
        log.warn(`Database '${name}' not found`);
      }
    } catch (error) {
      log.error(`Connection failed`, error);
    }
  }

  public async transact<T>(f: (transaction: Transaction) => Promise<any>): Promise<T> {
    try {
      return await this.instance.transaction(f);
    } catch (error) {
      log.error(`Error in transaction: ${error}, rolling back`);
      throw error;
    }
  }
}

const modelsPath = `${__dirname}/../../modules/**/*.model.+(t|j)s`;

export const sequelizeConfig: SequelizeOptions = {
  dialect: 'postgres',
  // tslint:disable-next-line:no-console
  logging: ConfigToUse.getBoolean('API_DB_LOG') ? console.log : false,
  models: [modelsPath],
  pool: {
    max: 10,
    min: 0
  }
};

export const DatabaseConnection = new SequelizeConnection(ConfigToUse.getString('DB_URL'), sequelizeConfig);

export interface IQueryOptions {
  transaction?: Transaction;
}
```

then in the composition root of the app we're simply initializing with:

```
    DatabaseConnection.initialize();
```

**Comment by RobinBuschmann** [comment]
Thanks. And the error is thrown on the initialization or when trying to query data?

**Comment by jpollard-cs** [comment]
@RobinBuschmann it appears to happen when attempting to query data. It seems as though the models are not properly getting initialized. Also interesting to note that a test for the `VendorEmployeeModel` equivalent was failing with an error indicating `create` was not defined on the model.

**Comment by jpollard-cs** [comment]
Ok I figured it out, but don't entirely understand the issue. I realized the issue was only happening for tests and finally noticed that in a test factory for the `VendorEmployeeModel` apparently there was one place we were importing the model inconsistently with the rest of the code base. We export the `VendorEmployeeModel` as a default, but then in the `index.ts` file we do  `export { default as VendorEmployeeModel } from './VendorEmployeeModel';`. I had actually thought this may be the issue  at one point (from looking at related issues) and had attempted to fix it, but missed a spot in this test file. Any insight into why this happens?
